sw/device/lib/runtime/epmp.h

```diff
@@ -0,0 +1,550 @@
+  kEpmpEntry14,
+  kEpmpEntry15,
+} epmp_entry_t;
+
+/**
+ * EPMP entry count.
+ */
+typedef enum {
+  kEpmpEntryCount0,
+  kEpmpEntryCount1,
+  kEpmpEntryCount2,
+  kEpmpEntryCount3,
+  kEpmpEntryCount4,
+  kEpmpEntryCount5,
+  kEpmpEntryCount6,
+  kEpmpEntryCount7,
+  kEpmpEntryCount8,
+  kEpmpEntryCount9,
+  kEpmpEntryCount10,
+  kEpmpEntryCount11,
+  kEpmpEntryCount12,
+  kEpmpEntryCount13,
+  kEpmpEntryCount14,
+  kEpmpEntryCount15,
+  kEpmpEntryCount16,
+} epmp_entry_count_t;
+
+/**
+ * EPMP generic status codes.
+ *
+ * These error codes can be used by any function. However if a function
+ * requires additional status codes, it must define a set of status codes to
+ * be used exclusively by such function.
+ */
+typedef enum {
+  kEpmpOk = 0,
+  kEpmpError,
+  kEpmpBadArg,
+} epmp_result_t;
+
+/**
+ * Initialize the EPMP library.
+ *
+ * The current state of the EPMP CSRs will be read and stored inside
+ * library. This function must be called before using the library. The
+ * library can only be initialized once.
+ *
+ * Note: it may be desirable to check the state is as expected. This
+ * can be done using `epmp_get_state`.
+ */
+epmp_result_t epmp_init(void);
+
+/**
+ * Check that the actual EPMP state matches what the library expects
+ * it to be.
+ */
+epmp_result_t epmp_check(void);
+
+/**
+ * Start an EPMP transaction.
+ *
+ * A new transaction will be initialized to expect the given number of
+ * entries to be reconfigured. Only one transaction may be in progress
+ * at a time. Each entry may only be reconfigured once per transaction.
+ * Call `epmp_transaction_end` to finalize the transaction and update
+ * the EPMP control registers.
+ *
+ * @param count Number of entries to be configured.
+ * @returns `epmp_result_t`
+ */
+epmp_result_t epmp_transaction_start(epmp_entry_count_t count);
+
+/**
+ * Finish an EPMP transaction and update EPMP control registers.
+ *
+ * Check that the expected number of regions have been reconfigured and
+ * then update the EPMP control registers. Once this call returns the
+ * transaction will be complete and a new transaction should be
+ * started if further updates are required.
+ *
+ * This call will also perform checks equivalent to an `epmp_check`
+ * call before modifying any registers.
+ *
+ * Updates will occur in the following sequence:
+ *
+ *   1. pmpaddr0-pmpaddr15
+ *   2. pmpcfg0-pmpcfg3
+ *
+ * @param count Number of entries that should have been configured.
+ * @returns `epmp_result_t`
+ */
+epmp_result_t epmp_transaction_end(epmp_entry_count_t count);
+
+/**
+ * EPMP configuration status codes.
+ */
+typedef enum {
+  kEpmpEntryConfigureOk = kEpmpOk,
+  kEpmpEntryConfigureError = kEpmpError,
+  kEpmpEntryConfigureBadArg = kEpmpBadArg,
+
+  /**
+   * Invalid addresses provided for the selected address mode.
+   */
+  kEpmpEntryConfigureBadRegion,
+
+  /**
+   * The transaction is in a bad state. This can be caused by:
+   *  * Missing call the `epmp_transaction_start`.
```

Comment:
maybe worth using `-` instead of `*` for sub-items to distinguish it from the main comment body?